c++/src/capnp/compat/byte-stream.c++

```diff
@@ -986,6 +981,45 @@ private:
     }
     KJ_UNREACHABLE;
   }
+
+  struct SplitResult {
+    kj::Array<kj::ArrayPtr<const byte>> piecesToWrite;
+    kj::Array<kj::ArrayPtr<const byte>> rest;
+  };
+
+  SplitResult splitPieces(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces, size_t limit) {
+    size_t splitByte = limit;
+    size_t splitPiece = 0;
+    while (pieces[splitPiece].size() <= splitByte) {
+      splitByte -= pieces[splitPiece].size();
+      ++splitPiece;
+    }
+
+    if (splitByte == 0) {
+      // Oh thank god, the split is between two pieces.
+      return SplitResult {
+        .piecesToWrite = kj::heapArray(pieces.slice(0, splitPiece)),
```

Comment:
This case didn't used to do allocation.

I was thinking something like:

```cpp
template <typename WritePieces>
kj::Promise<void> splitAndWrite(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces, size_t limit, WritePieces&& writeFirstPieces)
```

This would then call `writeFirstPieces(firstPieces)`, and then recurse to `write()` for the rest, and would handle all of the attaching of arrays to promises that is needed.

But if you're getting tired of this PR I guess put it back to the last version that had the duplicated code and we'll clean in up later.

Reply:
I think that makes a lot more sense, I pushed a fixup but could use some eyes.